---
title: "dba.plotProfile Demo"
author: "Rory Stark"
date: "14/04/2021"
output:
  html_document: 
    toc: yes
    fig_width: 7.5
    fig_height: 8
    number_sections: yes
  html_notebook: 
    toc: yes
    fig_width: 7
    fig_height: 9
  pdf_document: 
    fig_width: 7.5
    fig_height: 6
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(connectionObserver = NULL)
knitr::opts_knit$set(root.dir = "/data/personal/stark01/DiffBind_Vignette/")
library(DiffBind)
```

setwd("\~/Work/DiffBind/Vignette/DiffBind_Vignette/")
setwd("/data/personal/stark01/DiffBind_Vignette/")


# dba.plotProfile() Tests/Demo

Demonstration of `dba.plotProfile()`. This also serves as a test script for the function.

Due to the computational cost of this function, it is advised that the calculation of profiles and the plotting be separated into two calls, so that the profiles do not need to be re-generated if something goes wrong in the plotting. By default, when a `DBA` object is passed in to generate profiles, plotting is turned off and a `profileplyr` object is returned. When `dba.plotProfile()` is called with a `profileplyr` object, a plot if generated by default.

The main aspects of the profile plot are which **samples** are plotted (the X-axis) and which **sites** are plotted (the Y-axis). These can be specified in a number of flexible ways. Other parameters to `dba.plotProfile()` determine how the data are treated, controlling aspects such as how many sites are included in the plot, data normalization, sample merging (computing mean profiles for groups of samples), whether to annotate the sites based on genomic features, and control over the appearance of the plot.

## Default plots

The default plot depend on whether or not an analysis has been completed.

### Default plot: no analysis

If no analysis has been completed, the default plot will include all samples, in a single group. They will be merged based on the `DBA_REPLICATE` attribute, such that each sample class will have one heatmap based on the normalized mean read counts for all the replicate samples in that class.

Up to 1,000 of the consensus sites (randomly sampled) will be included, in a single group. If the genome is supported, annotation to nearby genomic features (promoters, genes, intragenic) will be determined and plotted.

```{r default_noanalysis, cache=TRUE}
data(tamoxifen_counts)
tamoxifen$config$RunParallel <- TRUE
profiles <- dba.plotProfile(tamoxifen)
dba.plotProfile(profiles)
```

### Default plot: analysis

If an analysis has been completed, the default will be based on the results of the first contrast. If the contrast compares two sample groups, the samples in those groups will be included, with the heatmaps colored separately for each side of the contrast. Samples are based on the `DBA_REPLICATE` attribute, such that each sample class will have one heatmap based on the normalized mean read counts for all the replicate samples in that class.

Only differentially bound sites are included, in two groups: Gain sites (with positive fold change) and Loss sites (negative fold change). If there are more than 1,000 sites in either category, the 1,000 sites with the great absolute value fold change will be included.

```{r default_analysis, cache=TRUE}
data(tamoxifen_analysis)
tamoxifen$config$RunParallel <- TRUE
profiles <- dba.plotProfile(tamoxifen)
dba.plotProfile(profiles)
```

### Default plot: merging cell types

In the sample experiment, there are multiple sample groups within each side of the contrast (The Resistant class has two sample groups based on the BT474 and MCF7 cell lines, while the Responsive class has three group, based on the MCF7, T47D, and ZR75 cell lines). If we want to generate profiles for the Responsive and Resistant classes, the `DBA_TISSUE` attribute can be added to the merge specification. By specifying`merge=c(DBA_TISSUE ,DBA_REPLICATE),` the mean counts (normalized) for all of the Resistant samples will calculated, as well as for all of the Responsive samples:

```{r default_merge_tissue, cache=TRUE}
profiles <- dba.plotProfile(tamoxifen,merge=c(DBA_TISSUE, DBA_REPLICATE))
dba.plotProfile(profiles)
```

## Specifying samples

The `samples` parameter can be used to specify a subset of samples to include in the plot. For example, we can limit the plot to only the MCF7 samples:

```{r default_samples_MCF7, cache=TRUE}
profiles <- dba.plotProfile(tamoxifen,samples=tamoxifen$mask$MCF7,
                            labels=c("MCF7_Responsive","MCF7_Resistant"))
dba.plotProfile(profiles)
```

The resulting plot shows two sample groups: Resistant and Responsive. This is because a) by default, the plots are based on the first contrast, b) by default, all samples in a group that differ only in their `DBA_REPLICATE` values ar merged, and c) by default, the label for a merged sample group includes only the attribute values that make it unique. As a result, this plot is showing only Gain/Loss sites for the first (Resistant vs Responsive) contrast, but *only using data for the MCF7 samples in the profiles*.

To see all of the MCF7 samples, specify `merge=NULL`, which will prevent the replicates from being merged:

```{r default_samples_MCF7_nomerge, cache=TRUE}
profiles <- dba.plotProfile(tamoxifen, samples=tamoxifen$mask$MCF7, merge=NULL)
dba.plotProfile(profiles)
```

In this plot, the sample group information from the contrast (Resistant/Responsive) has been lost, and the profile heatmaps are all plotted using the same color (red). This is because the specification for `samples` is a single vector. By specifying the samples as a list of two sample masks, the Resistant and Responsive MCF7 samples can be plotted using different colors:

```{r default_samples_MCF7_list, cache=TRUE}
profiles <- dba.plotProfile(tamoxifen,
                           samples=list(MCF7_Resistant=8:9,MCF7_Responsive=3:5),
                           merge=NULL)
dba.plotProfile(profiles)
```

## Specifying sites

There are a number of ways to specify which sites (peaks/genomic intervals) to include in the plot using the `sites` parameter.

A set of intervals in the form of a `GRanges` object can be passed directly. For example, the `dba.report()` function can be used to obtain a set of differentially bound sites, then subsetted to include those with a log fold change greater than 2, sorted from greatest positive fold change to greatest negative fold change:

```{r default_sites_granges, cache=TRUE}
rep <- dba.report(tamoxifen)
rep <- rep[abs(rep$Fold) > 2,]
rep <- rep[order(rep$Fold, decreasing=TRUE),]
profiles <- dba.plotProfile(tamoxifen, sites=rep,
                            samples=list(Resistant=tamoxifen$mask$Resistant,
                                         Responsive=tamoxifen$mask$Responsive))
dba.plotProfile(profiles)
```

Alternatively, a `GRangesList` object can be used to specify groups of sites, in this case separating the Gain and Loss sites:

```{r default_sites_granges_list, cache=TRUE}
repList <- GRangesList(Gain=rep[rep$Fold>0,],Loss=rep[rep$Fold<0,])
profiles <- dba.plotProfile(tamoxifen, sites=repList,
                            samples=list(Resistant=tamoxifen$mask$Resistant,
                                         Responsive=tamoxifen$mask$Responsive),
                            labels=list(sites=c("Resist","Respond")),
                            score="Fold")
dba.plotProfile(profiles)
```

Another way to specify sets of sites to profile is to use a report-based DBA object. This can be done implicitly, using a contrast, or explicitly, using the `dba.report()` function. We have already seen the implicit use in the default case where the first contrast is used (`sites=1`). If there are multiple contrasts, an alternative one may be selected:

```{r default_sites_contrast, cache=TRUE}
tamoxifen <- dba.contrast(tamoxifen,contrast=c("Tissue","ZR75","BT474"))
tamoxifen <- dba.analyze(tamoxifen)
profiles  <- dba.plotProfile(tamoxifen, sites=2)
dba.plotProfile(profiles)
```

The report-based object can also be passed in explicitly:

```{r default_sites_report, cache=TRUE}
repObj <- dba.report(tamoxifen, contrast=2, 
                     bDB=TRUE, bGain=TRUE, bLoss=TRUE,bAll=FALSE)
profiles <- dba.plotProfile(tamoxifen, sites=repObj,
                            samples=list(ZR75=tamoxifen$mask$ZR75,
                                         BT474=tamoxifen$mask$BT474))
dba.plotProfile(profiles)
```

More complex report-based objects can be used as well. For example we can plot the differentially bound sites from each of the two contrasts across all the samples:

```{r default_sites_report_2, cache=TRUE}
repObj <- dba.report(tamoxifen, contrast=1:2, bDB=TRUE)
profiles <- dba.plotProfile(tamoxifen, sites=repObj)
dba.plotProfile(profiles)
```

We can further refine the sites by considering the overlap between the differentially bound sites identified in the two contrasts:

```{r default_sites_report_venn, cache=TRUE}
overlaps <- dba.plotVenn(repObj,1:2)
```

Now we can generate separate profiles for the sites common to both contrasts, as well as those unique to one of them:

```{r default_sites_report_complex, cache=TRUE}
names(overlaps) <- c("Resistant-Responsive","ZR75-BT474","Both")
profiles <- dba.plotProfile(tamoxifen, sites=overlaps,
                            samples=list(ZR75=tamoxifen$masks$ZR75,
                                         BT474=tamoxifen$masks$BT474))
dba.plotProfile(profiles)

```

## Normalization

By default, the read counts in the profiles are adjusted by dividing by the normalization factors. This can be disabled by setting `normalize=FALSE`:

```{r default_sites_normalize_no, cache=TRUE}
profiles <- dba.plotProfile(tamoxifen, normalize=FALSE)
dba.plotProfile(profiles)

```

Compare with the second plot.

ADD NORMALIZATION VECTOR EXAMPLE HERE

## Changing profiling parameters

Parameters the control the profiling computations can be supplied to `dba.plotProfile()`. These are a subset of the parameters associated with the `profileplyer::BamBigwig_to_chipProfile()` function.

The default profile `style` is "point", which counts reads overlapping fixed-size bins around each site. The default `bin_size` is 20 base pairs, and the default `distanceAround` value, which controls how many base pairs up- and down-stream of each site to profile, is 1500 base pairs. This means that each row in the profile plot comprises 3000 base pairs, divided into 150 bins (each being 20bp).

The alternative way to specify the profiling `style` is "percentOfRegion", where the bin size is set by dividing the width of the sites to be plotted by the value of the `nOfWindows` parameter. In this case `distanceAround` represents the percentage of the site length to plot (up- and down- stream). For example, if `distanceAround=300`, then the profiled region will extend 3x the width of each profiled site, for a total of seven site widths (3 downstream, the site, and 3 upstream). In the example data set, where the peak widths are standardized to 400bp, if `nOfWindows=20`, then the bin size is 400/20=20bp. If `distanceAround=300`, then the region to be profiled will be extended 300% of the 400bp, or 1200bp up and downstream, for a total of 1200+1200+400=2800bp, divided into seven sets of 20bp bins, for a total of 140 bins to be profiled:

```{r default_sites_percentorregion, cache=TRUE}
profiles <-  dba.plotProfile(tamoxifen, style="percentOfRegion", 
                             nOfWindows=20, distanceAround=300)
dba.plotProfile(profiles)
```
